### PR TITLE
[c++][fix] Support Quantized Training with Categorical Features on CPU

### LINKED ISSUE #6257: [python-package] [bug] Segfault when combining `use_quantized_grad` with categorical features
## Description
LightGBM encounters a segfault when quantized gradients are used in combination with native categorical features.

## Reproducible examples
Training API Example:
```
import numpy as np
import lightgbm as lgbm

rng = np.random.default_rng(1)
X = rng.choice([1, 2], size=(10_000, 1))
y = rng.choice([0, 1], size=(10_000,))

train_set = lgbm.Dataset(
    data=X,
    feature_name=["A"],
    categorical_feature=["A"],
    label=y,
    params={"random_seed": 1},
    free_raw_data=False,
).construct()

booster = lgbm.train(
    params={
        "objective": "binary",
        "use_quantized_grad": True,
        "random_state": 1,
    },
    num_boost_round=10,
    train_set=train_set,
)
```

Scikit-learn API Example:
```
import numpy as np
import lightgbm as lgbm

rng = np.random.default_rng(1)
X = rng.choice([1, 2], size=(10_000, 1))
y = rng.choice([0, 1], size=(10_000,))

clf = lgbm.LGBMClassifier(
    objective="binary",
    use_quantized_grad=True,
    random_state=1,
    num_boost_round=10,
)
clf.fit(
    X=X,
    y=y,
    feature_name=["A"],
    categorical_feature=["A"],
)
```

Both examples result in a core dump when LightGBM attempts to fit the booster. The failure occurs for both regression and classification objectives. The failure also occurs regardless of the number of features as long as at least one feature is categorical. No error is encountered when either `use_quantized_grad=False` or no categorical features are specified.

## Environment info

OS: Ubuntu 22.04.3 LTS (Jammy Jellyfish)
Python Version: 3.10.8
Relevant python packages:
```
lightgbm==4.2.0
numpy==1.26.0
scikit-learn==1.3.2
scipy==1.11.4
```

`lightgbm` was built from source via `pip install --no-binary lightgbm lightgbm` using GCC=11.4.0. I also attempted to install the pre-compiled version via PyPI (`pip install lightgbm`) and encountered the same segfault.

### PR DESCRIPTION
**Pull request by shiyu1994**
This is to partially fix #6257. Currently quantized training is not supported with categorical features. This PR add the support on CPUs.

### CODE REVIEW

**Comment by jameslamb** [comment]
I believe a PR can only have one of the labels recognized by this project's `release-drafter` configuration at https://github.com/microsoft/LightGBM/blob/master/.github/release-drafter.yml. I removed `fix` because I think this is more of a `feature` (even if we initially found out about it because of a bug report).

**Comment by shiyu1994** [comment]
@jameslamb Thanks for the correction.

**Comment by shiyu1994** [comment]
@guolinke Could you help to review this when you have time? Thanks.

File: src/treelearner/feature_histogram.cpp
Comment (guolinke) [inline src/treelearner/feature_histogram.cpp:1]: this is from hpp file?

Comment (shiyu1994) [reply src/treelearner/feature_histogram.cpp:1]: Yes, it is separated to avoid too large obj file, which causes the failure of this job.
https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=15803&view=logs&j=1f4df553-f999-5fff-c6fe-71123c872ae2
